File: oolab/src/main/java/agh/ics/oop/model/Animal.java
package agh.ics.oop.model;

import java.util.Objects;

public class Animal {
    static final Vector2d lowerMapBoundary = new Vector2d(0, 0);
    static final Vector2d upperMapBoundary = new Vector2d(4, 4);
    private MapDirection facingDirection;
    private Vector2d localizationOnMap;

    public Animal() {
        facingDirection = MapDirection.NORTH;
        localizationOnMap = new Vector2d(2, 2);
    }

    public Animal(Vector2d localizationOnMap) {
        this.localizationOnMap = localizationOnMap;
        facingDirection = MapDirection.NORTH;
    }

    public Vector2d getLocalizationOnMap() {
        return localizationOnMap;
    }

    public MapDirection getFacingDirection() {
        return facingDirection;
    }

    public void move(MoveDirection direction) {
        switch (direction) {
            case LEFT -> facingDirection = facingDirection.previous();
            case RIGHT -> facingDirection = facingDirection.next();
            case FORWARD -> {
                var new_loc = this.getLocalizationOnMap().add(this.facingDirection.toUnitVector());
                if (new_loc.follows(lowerMapBoundary)
                        && new_loc.precedes(upperMapBoundary))
                    localizationOnMap = new_loc;
            }
            case BACKWARD -> {
                var new_loc = this.getLocalizationOnMap().subtract(this.facingDirection.toUnitVector());
                if (new_loc.follows(lowerMapBoundary)
                        && new_loc.precedes(upperMapBoundary))
                    localizationOnMap = new_loc;
            }
        }
    }

    @Override
    public String toString() {
        return String.format("(position: %s , orientation %s)", localizationOnMap, facingDirection);
    }

    public boolean isAt(Vector2d position) {
        return Objects.equals(localizationOnMap, position);
    }
}

File: oolab_kotlin/src/main/kotlin/Animal.kt
package org.example

import java.util.*

class Animal( var localizationOnMap: Vector2d = Vector2d(2, 2), val id:Int){
    var facingDirection: MapDirection = MapDirection.NORTH;

    fun move(direction: MoveDirection, validator: MoveValidator){
        when(direction) {
            MoveDirection.LEFT -> facingDirection = facingDirection.previous();
            MoveDirection.RIGHT -> facingDirection = facingDirection.next();
            MoveDirection.FORWARD -> {
                var newLoc = localizationOnMap + facingDirection.vector2d;
                if(validator.canMoveTo(newLoc)){
                    localizationOnMap = newLoc;
                }
            }
            MoveDirection.BACKWARD ->  {
                var newLoc = localizationOnMap - facingDirection.vector2d;
                if(validator.canMoveTo(newLoc)){
                    localizationOnMap = newLoc;
                }
            }
        }
    }

    override fun toString(): String {
        return return String.format(
            when (facingDirection) {
                MapDirection.NORTH -> "^"
                MapDirection.EAST -> ">"
                MapDirection.WEST -> "<"
                MapDirection.SOUTH -> "v"
            }
        )
    }

    fun isAt(position: Vector2d): Boolean {
        return localizationOnMap == position;
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Animal

        if (localizationOnMap != other.localizationOnMap) return false
        if (facingDirection != other.facingDirection) return false
        if (id != other.id) return false

        return true
    }

    override fun hashCode(): Int {
        return Objects.hash(localizationOnMap, facingDirection, id);
    }


}